[Driver] SR-2396: Driver should have a -verify-debug-info option

matthewcarroll opened the pull request:
Add a `-verify-debug-info` option that invokes `dwarfdump --verify` as the last step after running `dsymutil`. `dwarfdump` is invoked with same options clang 802.0.27.2 uses to invoke it:
`dwarfdump --verify --debug-info --eh-frame --quiet`

A warning is produced if `-verify-debug-info` is set and no debug option is also set.

`dwarfdump` is failing to validate the debug info in the test `verify-debug-info.swift`. The failure is:
`error: .debug_line[0x0000007d].row[0].file = 1 is not a valid index`

What addtional tests need to be added?
Would you like to add additional options to the driver provided by dwarfdump?

https://bugs.swift.org/browse/SR-2396

adrian-prantl commented on include/swift/AST/DiagnosticsDriver.def:
```diff
@@ -121,6 +121,9 @@ ERROR(error_conflicting_options, none,
       (StringRef, StringRef))
 
+WARNING(verify_debug_info_requires_debug_option,none,
+        "ignoring '-verify-debug-info'; no debug option was set", ())
```
ignoring '-verify-debug-info'; no debug *info* option was set

adrian-prantl replied:
This still looks odd when using -gnone.
What about:
ignoring '-verify-debug-info'; no debug info is being generated

adrian-prantl commented on include/swift/Option/Options.td:
```diff
@@ -342,6 +342,12 @@ def gdwarf_types : Flag<["-"], "gdwarf-types">,
   HelpText<"Emit full DWARF type info.">;
 
+// Verify debug info
+
```
remove extra newline

adrian-prantl commented on lib/Driver/Driver.cpp:
```diff
@@ -156,6 +156,19 @@ static void validateArgs(DiagnosticEngine &diags, const ArgList &Args) {
                    "-warnings-as-errors", "-suppress-warnings");
   }
+  
+  // Check for missing debug option when verifying debug info
```
missing '.' at the end

adrian-prantl commented on lib/Driver/Driver.cpp:
```diff
@@ -156,6 +156,19 @@ static void validateArgs(DiagnosticEngine &diags, const ArgList &Args) {
+  if (Args.hasArg(options::OPT_verify_debug_info)) {
+    bool hasDebugOption = true;
+    Arg *Arg = Args.getLastArg(swift::options::OPT_g_Group);
+    if (!Arg || Arg->getOption().matches(swift::options::OPT_gnone)) {
```
remove extra brace

adrian-prantl commented on lib/Driver/Driver.cpp:
```diff
@@ -156,6 +156,19 @@ static void validateArgs(DiagnosticEngine &diags, const ArgList &Args) {
+    if (!Arg || Arg->getOption().matches(swift::options::OPT_gnone)) {
+      hasDebugOption = false;
+    }
+    if (!hasDebugOption) {
```
remove extra brace

adrian-prantl commented on lib/Driver/ToolChains.cpp:
```diff
@@ -748,6 +748,25 @@ ToolChain::constructInvocation(const GenerateDSYMJobAction &job,
+                               const JobContext &context) const {
+  assert(context.Inputs.size() == 1);
+  assert(context.InputActions.empty());
+
```
Please add:
// This mirrors the clang driver's --verify-debug-info option.

adrian-prantl reviewed (approved):
Thanks, this LGTM with some minor feedback addressed!

adrian-prantl commented:
@swift-ci test and merge

adrian-prantl commented:
@swift-ci test and merge
test failure looks unrelated.

adrian-prantl commented:
@swift-ci test and merge

adrian-prantl commented:
@swift-ci test and merge

adrian-prantl commented:
TEST 'Swift(macosx-x86_64) :: Driver/actions.swift' FAILED ********************

Could you take a look?

matthewcarroll commented:
Yes, `dwarfdump` is failing to validate the debug info in the test verify-debug-info.swift. The failure is:
`error: .debug_line[0x0000007d].row[0].file = 1 is not a valid index`
I get the same result from the command line using `swiftc` in Xcode 8.3 beta and `dwarfdump` 138.

adrian-prantl commented:
Sweet, we found a bug! Could you just change test test to only verify the driver decisions (check the output of -###) without actually executing it?

slavapestov commented:
@swift-ci Please test

matthewcarroll commented:
@adrian-prantl,

I think this is ready. Would you mind retesting it?

Thanks!

adrian-prantl commented:
@swift-ci please test and merge

matthewcarroll committed:
[Driver] SR-2396: Driver should have a -verify-debug-info option

Add a -verify-debug-info option that invokes dwarfdump --verify as the last step after running dsymutil. dwarfdump is invoked with same options clang 802.0.35 uses to invoke it:

dwarfdump --verify --debug-info --eh-frame --quiet

A warning is produced if -verify-debug-info is set and no debug option is also set.

dwarfdump is failing to validate the debug info in the test verify-debug-info.swift. The failure is:
error: .debug_line[0x0000007d].row[0].file = 1 is not a valid index

https://bugs.swift.org/browse/SR-2396

matthewcarroll commented:
@adrian-prantl,

Would you mind retesting this?

adrian-prantl commented:
@swift-ci please test

matthewcarroll commented:
@adrian-prantl,

Thanks. All tests passed.
